src/platform/storage_winuser.c

```diff
@@ -121,38 +122,50 @@ CxPlatStorageOpen(
     }
 
     CxPlatZeroMemory(Storage, sizeof(CXPLAT_STORAGE));
-    Storage->Callback = Callback;
-    Storage->CallbackContext = CallbackContext;
+    if (Callback != NULL) {
+        Storage->Callback = Callback;
+        Storage->CallbackContext = CallbackContext;
 
-    Storage->NotifyEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
-    if (Storage->NotifyEvent == NULL) {
-        Status = QUIC_STATUS_OUT_OF_MEMORY;
-        goto Exit;
-    }
+        Storage->NotifyEvent = CreateEvent(NULL, FALSE, FALSE, NULL);
+        if (Storage->NotifyEvent == NULL) {
+            Status = QUIC_STATUS_OUT_OF_MEMORY;
+            goto Exit;
+        }
 
-    Storage->ThreadPoolWait =
-        CreateThreadpoolWait(
-            CxPlatStorageRegKeyChangeCallback,
-            Storage,
-            NULL);
-    if (Storage->ThreadPoolWait == NULL) {
-        Status = QUIC_STATUS_OUT_OF_MEMORY;
-        goto Exit;
+        Storage->ThreadPoolWait =
+            CreateThreadpoolWait(
+                CxPlatStorageRegKeyChangeCallback,
+                Storage,
+                NULL);
+        if (Storage->ThreadPoolWait == NULL) {
+            Status = QUIC_STATUS_OUT_OF_MEMORY;
+            goto Exit;
+        }
     }
 
     QuicTraceLogVerbose(
         StorageOpenKey,
         "[ reg] Opening %s",
         FullKeyName);
 
+    REGSAM DesiredAccess = KEY_READ | KEY_NOTIFY;
+
+    if (Flags & CXPLAT_STORAGE_OPEN_FLAG_WRITABLE) {
+        DesiredAccess |= KEY_WRITE;
+    }
+
+    if (Flags & CXPLAT_STORAGE_OPEN_FLAG_DELETEABLE) {
+        DesiredAccess |= DELETE;
+    }
+
 #pragma prefast(suppress:6001, "SAL can't track FullKeyName")
     Status =
         HRESULT_FROM_WIN32(
         RegOpenKeyExA(
```

Comment:
If the key doesn't exist yet, it might have to be created (only when the writable flag is set)